CARDS-2723: Create an endpoint to dump data processors and filters

acrowthe opened the pull request:
To view the results of this endpoint, navigate to http://localhost:8080/Questionnaires.selectors.md for Markdown or http://localhost:8080/Questionnaires.selectors.json for json.

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
- Base implementation

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
- Add descriptions to all processors and filters

acrowthe commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/SimpleAnswerLabelProcessor.java:
```diff
@@ -58,6 +58,13 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Only runs on forms\n."
```
```suggestion
        return "Only runs on forms.\n"
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/BareFormProcessor.java:
```diff
@@ -58,6 +58,14 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Only runs on forms.\n"
```
I would capitalize the initial of resource name and maybe also format it a bit differently to make it stand out.
```suggestion
        return "Only runs on `Forms`.\n"
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/ExcludeFilesProcessor.java:
```diff
@@ -59,6 +60,20 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Exclude file contents from an export. By default, exclude all files";
```
```suggestion
        return "Exclude the contents of uploaded files from the serialization. By default, all uploaded files are excluded.";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/ExcludeFilesProcessor.java:
```diff
@@ -59,6 +60,20 @@ public String getName()
+    {
+        return new SelectorDetails(getName(), getDescription(), isEnabledByDefault(null),
+             "exclude", "If this option is included, only exclude the specified file. "
+            + "`.excludeFiles:exclude=/Questionnaires/Path/To/Question`");
```
```suggestion
            + "`.excludeFiles:exclude=/Questionnaires/Path/To/QuestionWithFileAnswer`");
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/FormAnswerCopyProcessor.java:
```diff
@@ -56,6 +56,17 @@ public class FormAnswerCopyProcessor extends AbstractAnswerCopyProcessor
+        return "Only runs on Forms.\n"
+            + "Copy the answer of certain answers to the form JSON. The answers to copy are configured in "
+            + "`/apps/cards/config/CopyAnswers/Questionnaires/[questionnaire name]/` as properties with the desired "
+            + "output name as the key and references to a questiona s the value.\n"
```
```suggestion
            + "output name as the key and references to a question as the value.\n"
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/FormAnswerCopyProcessor.java:
```diff
@@ -56,6 +56,17 @@ public class FormAnswerCopyProcessor extends AbstractAnswerCopyProcessor
+            + "`/apps/cards/config/CopyAnswers/Questionnaires/[questionnaire name]/` as properties with the desired "
+            + "output name as the key and references to a questiona s the value.\n"
+            + "Questions can be copied either from the current form itself, or from another form belonging to the "
+            + "same subject.";
```
Or one of it's parents or one of its descendants? It gets complicated if we want to be accurate...

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/FormAnswerFilterProcessor.java:
```diff
@@ -73,6 +74,26 @@ public String getName()
+    {
+        return "Only runs on forms.\n"
+            + "Include or exclude answers or answer sections based on the question or section they refer to.\n"
+            + "If only include options are provided, only those nodes and any descendants will be included\n"
+            + "If only exclude options are provided, all other nodes will be included.\n"
```
```suggestion
            + "If only `include` options are provided, only those items and any descendants will be included\n"
            + "If only `exclude` options are provided, all other items will be included.\n"
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/FormAnswerFilterProcessor.java:
```diff
@@ -73,6 +74,26 @@ public String getName()
+            + "Include or exclude answers or answer sections based on the question or section they refer to.\n"
+            + "If only include options are provided, only those nodes and any descendants will be included\n"
+            + "If only exclude options are provided, all other nodes will be included.\n"
+            + "If both include and exclude options are provided, then for a node to be included it must be a "
+            + "specified include node or descendent thereof and also not be a member or descendant of an exclude node."
+            + "\nIt is not possible to include a descendant of an excluded node";
```
```suggestion
            + "If both `include` and `exclude` options are provided, then for an item to be included it must be "
            + "listed as an `include` item or descendent thereof and also not be listed as or descendant of an `exclude` item."
            + "\nIt is not possible to include a descendant of an excluded item.";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/FormAnswerFilterProcessor.java:
```diff
@@ -73,6 +74,26 @@ public String getName()
+    public SelectorDetails getDetails()
+    {
+        return new SelectorDetails(getName(), getDescription(), isEnabledByDefault(null),
+            "include", "A path to an included node. `answerFilter:include=/Questionnaires/Path/To/Question`",
+            "exclude", "A path to an excluded node. answerFilter:exclude=/Questionnaires/Path/To/Question");
```
To check of the lines aren't too long for the checkstyle plugin:
```suggestion
            "include", "A path to an included question or section. `answerFilter:include=/Questionnaires/Path/To/Question`",
            "exclude", "A path to an excluded question or section. answerFilter:exclude=/Questionnaires/Path/To/Question");
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireFilterProcessor.java:
```diff
@@ -67,6 +68,26 @@ public boolean canProcess(final Resource resource)
+    @Override
+    public String getDescription()
+    {
+        return "Only runs on Questionnaires.\n"
+            + "Include or exclude question or sections.\n"
+            + "If only include options are provided, only those nodes and any descendants will be included\n"
+            + "If only exclude options are provided, all other nodes will be included.\n"
+            + "If both include and exclude options are provided, then for a node to be included it must be a "
+            + "specified include node or descendent thereof and also not be a member or descendant of an exclude node."
+            + "\nIt is not possible to include a descendant of an excluded node";
```
Similar rewording to the description for Forms

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireFilterProcessor.java:
```diff
@@ -67,6 +68,26 @@ public boolean canProcess(final Resource resource)
+    public SelectorDetails getDetails()
+    {
+        return new SelectorDetails(getName(), getDescription(), isEnabledByDefault(null),
+            "include", "A path to an included node. `answerFilter:include=/Questionnaires/Path/To/Question`",
+            "exclude", "A path to an excluded node. answerFilter:exclude=/Questionnaires/Path/To/Question");
```
Similar rewording to the details for Forms

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/SubjectAnswerCopyProcessor.java:
```diff
@@ -60,6 +60,16 @@ public class SubjectAnswerCopyProcessor extends AbstractAnswerCopyProcessor
+    public String getDescription()
+    {
+        return "Only runs on Subjects.\n"
+            + "Copy the answer of certain answers from forms to the root subject JSON. The answers to copy are "
+            + "configured in `/apps/cards/config/CopyAnswers/Questionnaires/[questionnaire name]/` as properties with "
+            + "the desired output name as the key and references to a questiona s the value.\n"
+            + "Questions can be copied either from a form belonging to this subject, or one of it's ancestors.";
```
```suggestion
            + "Copy the valueof certain answers from forms to the root subject JSON. The answers to copy are "
            + "configured in `/apps/cards/config/CopyAnswers/Questionnaires/[questionnaire name]/` as properties with "
            + "the desired output name as the key and references to a question as the value.\n"
            + "Questions can be copied either from a form belonging to this subject, or one of it's ancestors.";
```

marta- replied:
Same clarification as for Forms needed here.

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/FormAnswerCopyProcessor.java:
```diff
@@ -56,6 +56,17 @@ public class FormAnswerCopyProcessor extends AbstractAnswerCopyProcessor
+    public String getDescription()
+    {
+        return "Only runs on Forms.\n"
+            + "Copy the answer of certain answers to the form JSON. The answers to copy are configured in "
```
```suggestion
            + "Copy the values of certain answers to the form JSON. The answers to copy are configured in "
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/SubjectAnswerCopyProcessor.java:
```diff
@@ -60,6 +60,16 @@ public class SubjectAnswerCopyProcessor extends AbstractAnswerCopyProcessor
+    public String getDescription()
+    {
+        return "Only runs on Subjects.\n"
+            + "Copy the answer of certain answers from forms to the root subject JSON. The answers to copy are "
```
```suggestion
            + "Copy the values of certain answers from forms to the root subject JSON. The answers to copy are "
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/SubjectAnswerCopyProcessor.java:
```diff
@@ -60,6 +60,16 @@ public class SubjectAnswerCopyProcessor extends AbstractAnswerCopyProcessor
+        return "Only runs on Subjects.\n"
+            + "Copy the answer of certain answers from forms to the root subject JSON. The answers to copy are "
+            + "configured in `/apps/cards/config/CopyAnswers/Questionnaires/[questionnaire name]/` as properties with "
+            + "the desired output name as the key and references to a questiona s the value.\n"
```
```suggestion
            + "the desired output name as the key and references to a question as the value.\n"
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/CreatedAfter.java:
```diff
@@ -39,6 +41,17 @@
+        List<SelectorDetails> result = new ArrayList<>();
+        result.add(new SelectorDetails("createdAfter",
+            "Only show results that were created after the requested time. "
+            + "eg. `.dataFilter:createdAfter=2025-01-01T06:00:00%5C.000-05:00` for forms created after "
```
```suggestion
            + "e.g. `.dataFilter:createdAfter=2025-01-01T06:00:00%5C.000-05:00` for forms created after "
```

marta- replied:
Is using all the subdivisions required or would the short format "2025-01-01" also work? And if yes, would that mean end of the day or start of the day?

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/CreatedAfter.java:
```diff
@@ -39,6 +41,17 @@
+    {
+        List<SelectorDetails> result = new ArrayList<>();
+        result.add(new SelectorDetails("createdAfter",
+            "Only show results that were created after the requested time. "
```
```suggestion
            "Only show results that were created after the requested datetime. "
            + "The datetime must be specified in ISO 8601 format. "
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/CreatedBefore.java:
```diff
@@ -39,6 +41,17 @@
+        result.add(new SelectorDetails("createdBefore",
+            "Only show results that were created before the requested time. "
+            + "eg. `.dataFilter:createdBefore=2025-01-01T06:00:00%5C.000-05:00` for forms created before "
+            + "January 1, 2025 at 6 AM in the time zone UTC-5"));
```
Same comments as for createdAfter

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/Creator.java:
```diff
@@ -41,6 +43,17 @@
+        result.add(new SelectorDetails("createdBy",
+            "Only show results that were created by the specified user."));
+        result.add(new SelectorDetails("notCreatedBy",
+            "Only show results that were not created by the specified user."));
```
"not" can easily go unnoticed (at least by me)
```suggestion
            "Only show results that were created by any user other than the specified user."));
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/LastAuthor.java:
```diff
@@ -41,6 +43,17 @@
+        result.add(new SelectorDetails("modifiedBy",
+            "Only show results that were last modified by the specified user."));
+        result.add(new SelectorDetails("notModifiedBy",
+            "Only show results that were not last modified by the specified user."));
```
```suggestion
            "Only show results that were last modified by any user other than the specified user."));
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/ModifiedAfter.java:
```diff
@@ -39,6 +41,17 @@
+        result.add(new SelectorDetails("modifiedAfter",
+            "Only show results that were modified after the requested time. "
+            + "eg. `.dataFilter:modifiedAfter=2025-01-01T06:00:00%5C.000-05:00` for forms modified after "
+            + "January 1, 2025 at 6 AM in the time zone UTC-5"));
```
Same comments as for createdBefore/After

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/ModifiedBefore.java:
```diff
@@ -39,6 +41,17 @@
+        result.add(new SelectorDetails("modifiedBefore",
+            "Only show results that were modified before the requested time. "
+            + "eg. `.dataFilter:modifiedBefore=2025-01-01T06:00:00%5C.000-05:00` for forms modified before "
+            + "January 1, 2025 at 6 AM in the time zone UTC-5"));
```
Here as well

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/StatusFlags.java:
```diff
@@ -41,6 +43,17 @@
+    {
+        List<SelectorDetails> result = new ArrayList<>();
+        result.add(new SelectorDetails("status",
+            "Only show results that include the specified status flag."));
```
```suggestion
            "Only show results that have the specified status flag."));
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/filters/StatusFlags.java:
```diff
@@ -41,6 +43,17 @@
+        result.add(new SelectorDetails("status",
+            "Only show results that include the specified status flag."));
+        result.add(new SelectorDetails("statusNot",
+            "Only show results that do not include the specified status flag."));
```
Maybe capitalizing NOT is enough to make it stand out and we could do the same for the previous filters that are similar?
```suggestion
            "Only show results that do NOT have the specified status flag."));
```

acrowthe replied:
I don't really like the idea of capitalized NOT.
However, `...that do not have the...` feels different enough compared to `...that have the...` that the extra emphasis is not required in this specific case

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/BooleanLabelProcessor.java:
```diff
@@ -45,6 +45,12 @@ public class BooleanLabelProcessor extends SimpleAnswerLabelProcessor implements
+    @Override
+    public String getDescription()
+    {
+        return super.getDescription() + "Get the human readable answer for boolean questions.";
```
```suggestion
        return super.getDescription() + "Get the human readable answer for boolean questions by outputting the labels specified in the question definition, e.g. "Yes" or "True" instead of "1".";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/DateLabelProcessor.java:
```diff
@@ -46,6 +46,12 @@ public class DateLabelProcessor extends SimpleAnswerLabelProcessor implements Re
+    @Override
+    public String getDescription()
+    {
+        return super.getDescription() + "Get the human readable answer for date questions.";
```
To be reformatted because the updated line is too long

```suggestion
        return super.getDescription() + "Get the human readable answer for date questions. The human readable version is the date formatted with the date format configured in the date question definition, for example `01/07/2025` instead of the stored value `2025-01-07T00:00:00%5C.000-05:00`.";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/AnswerOptionsLabelProcessor.java:
```diff
@@ -42,6 +42,12 @@
+    @Override
+    public String getDescription()
+    {
+        return super.getDescription() + "Get the human readable answer for text and number questions with options.";
```
```suggestion
        return super.getDescription() + "Get the human readable answer for text and number questions with options, by outputting the answer option's label instead of the stored value.";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/FileLabelProcessor.java:
```diff
@@ -42,6 +42,12 @@
+    @Override
+    public String getDescription()
+    {
+        return super.getDescription() + "Get the human readable answer for file questions.";
```
Is this what we're doing? If not, please provide the correct details.
```suggestion
        return super.getDescription() + "Get the human readable answer for file questions by outputting the file name instead of the path.";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/SimpleAnswerLabelProcessor.java:
```diff
@@ -58,6 +58,13 @@ public String getName()
+    public String getDescription()
+    {
+        return "Only runs on forms\n."
+            + "Add the human readable version of an answer to it's serialization in the `displayedValue` property";
```
```suggestion
            + "Add the human readable version of an answer to its serialization in the `displayedValue` property";
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/labels/VocabularyOptionsLabelProcessor.java:
```diff
@@ -42,6 +42,13 @@
+    public String getDescription()
+    {
+        return "Runs on Forms and Questionnaires.\n"
+            + "Adds a label to answer option nodes.";
```
This doesn't sound right, will come back to it unless someone else rephrases it first.

marta- commented on modules/data-model/links/impl/src/main/java/io/uhndata/cards/links/internal/serialize/LinksProcessor.java:
```diff
@@ -51,6 +51,12 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Simplify link serialization to only include the link type and path to the linked resource";
```
Do we support links to anything or just forms?
```suggestion
        return "Simplify the serialization of links to other resources (such as other forms) to only include the link type and path to the linked resource.";
```

marta- commented on modules/data-model/links/impl/src/main/java/io/uhndata/cards/links/internal/serialize/SkipLinksProcessor.java:
```diff
@@ -42,6 +42,12 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Skip all links";
```
```suggestion
        return "Exclude the links to other resources from serialization.";
```

marta- commented on modules/data-model/subjects/impl/src/main/java/io/uhndata/cards/subjects/internal/serialize/SimpleSubjectProcessor.java:
```diff
@@ -45,6 +45,13 @@ public String getName()
+    public String getDescription()
+    {
+        return "Only runs on Subjects.\n"
+            + "Simplify form serialization by removing child jcr properties.";
```
"child jcr properties" makes me think of "descendent nodes' jcr properties". If that's not what is intended, drop "child"
```suggestion
            + "Simplify form serialization by removing the jcr properties.";
```

marta- commented on modules/data-model/subjects/impl/src/main/java/io/uhndata/cards/subjects/internal/serialize/SubjectTypeInstanceCountProcessor.java:
```diff
@@ -56,6 +56,13 @@ public String getName()
+    public String getDescription()
+    {
+        return "Only runs on Subjects.\n"
+            + "Add the number of instances of a subject to the subject type json";
```
```suggestion
            + "Include the number of subjects of that type in the subject type json";
```

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/ClinicDataFilterFactory.java:
```diff
@@ -52,6 +54,19 @@ public class ClinicDataFilterFactory extends BaseFilterFactory implements DataFi
+    {
+        List<SelectorDetails> result = new ArrayList<>();
+        result.add(new SelectorDetails("clinic",
+            "Only show forms that belong to a user that has a visit information form for the specified clinic. "
```
```suggestion
            "Only show forms that belong to a user that has a "Visit information" form for the specified clinic. "
```

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/ClinicDataFilterFactory.java:
```diff
@@ -52,6 +54,19 @@ public class ClinicDataFilterFactory extends BaseFilterFactory implements DataFi
+            "Only show forms that belong to a user that has a visit information form for the specified clinic. "
+            + "If included multiple times, this includes forms belonging to any of the specified clinics."));
+        result.add(new SelectorDetails("clinicNot",
+            "Exclude forms that belong to a user that has a visit information form for the specified clinic. "
```
```suggestion
            "Exclude forms that belong to a user that has a "Visit information" form for the specified clinic. "
```

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/serialize/FormToSurveyLinkProcessor.java:
```diff
@@ -51,6 +51,13 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "If a node has a `belongsToSurvey` link, include the name of the linked form in the root of the form "
```
```suggestion
        return "If a form has a `belongsToSurvey` link, include the name of the linked form in the root of that form's serialization JSON."
```

veronikaslc commented on modules/utils/src/main/java/io/uhndata/cards/serialize/spi/SelectorDetails.java:
```diff
@@ -0,0 +1,135 @@
+package io.uhndata.cards.serialize.spi;
+
+/**
+ * The details about a selctor that can be used to filter or modify what data is returned.
```
```suggestion
 * The details about a selector that can be used to filter or modify what data is returned.
```

marta- commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/DeepProcessor.java:
```diff
@@ -41,6 +41,12 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Enable deep serialization of nodes, including all child nodes";
```
```suggestion
        return "Enable deep serialization, i.e. including the serialization of all descendant items. For example, serializing a Form without `deep` would only output the form's properties. With `deep`, all the answers are included, organized by sections.";
```

marta- commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/ExcludeDefaultPropertiesProcessor.java:
```diff
@@ -73,6 +73,15 @@ public String getName()
+    public String getDescription()
+    {
+        return "Exclude properties if their value is the default value or an obvious non-value:\n"
+            + " - If a properties' value matches it's JCR-defined default value\n"
```
```suggestion
            + " - If a property's value matches its JCR-defined default value\n"
```

marta- commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/IdentificationProcessor.java:
```diff
@@ -45,6 +45,12 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Identify a node by including it's `@path` and `@name` properties.";
```
```suggestion
        return "Identify a node by including its `@path` and `@name` properties.";
```

marta- commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/PropertiesProcessor.java:
```diff
@@ -48,6 +48,12 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Serialize all node properties.";
```
```suggestion
        return "Serialize all properties.";
```

marta- commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/ReferencedProcessor.java:
```diff
@@ -45,6 +45,12 @@ public String getName()
+    @Override
+    public String getDescription()
+    {
+        return "Report if a node is referenced by adding a `@referenced=true|false` property";
```
```suggestion
        return "Report if a resource is referenced by adding a `@referenced=true|false` property";
```

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
- Update endpoint to order JsonObject children by name

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
- Update descriptions to be more user friendly

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
- Update documentation endpoint headers to be one level deeper to better fit into desired markdown
- Create markdown readme file with endpoint contents and an additional explination

marta- committed:
CARDS-2723: Create an endpoint to dump data processors and filters

Updates to the markdown readme file:
* Added title
* Shifted headings of sections to a lower level to account for the addition of the title
* Removed duplicated content artifact

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+### Export formats
+Multiple different export formats are supported. These include:
+- **.csv** and **.tsv**: This format includes additional processing options to help match data into a row/column format. These options are explained in the CSV Adapter Options section below, alongside the standard Processor and Filter options.
+- **.json**: This format most closely matches the way that data is stored internally.
```
I would put JSON first.

```suggestion
- **.json**: This format most closely matches the way that data is stored internally. It is also highly customizable, from very verbose to the bare minimum.
- **.csv** and **.tsv**: This format includes additional processing options to help match data into a row/column format. These options are explained in the CSV Adapter Options section below, alongside the standard Processor and Filter options.
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- **.csv** and **.tsv**: This format includes additional processing options to help match data into a row/column format. These options are explained in the CSV Adapter Options section below, alongside the standard Processor and Filter options.
+- **.json**: This format most closely matches the way that data is stored internally.
+- **.txt**: minimalistic human-readable serialization, suitable for copy/pasting in a note
+- **.md**: suitable for printing or archiving
```
```suggestion
- **.md**: Markdown-formatted version of the text format, suitable for printing or archiving
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- `/Questionnaires/<QuestionnaireId>.deep.json` exports the full Questionnaire definition, including questions and sections
+- `/Questionnaires/<QuestionnaireId>.data.json` exports the forms containing answers to the specified Questionnaire. Filters and processors can be added, as shown in the examples above.
+- `/Subjects/<MRN>/<Encounter ID>.data.deep.json` exports a visit and all its associated forms. Filters and processors can be added, as shown in the examples above.
+- `/Forms/<Form ID>.deep.json` exports a Form with all its answers. Filters and processors can be added, as shown in the examples above.
```
Now the examples are below, not above.

```suggestion
- `/Questionnaires/<QuestionnaireId>.data.json` exports the forms containing answers to the specified Questionnaire. Filters and processors can be added, as shown in the examples below.
- `/Subjects/<MRN>/<Encounter ID>.data.deep.json` exports a visit and all its associated forms. Filters and processors can be added, as shown in the examples below.
- `/Forms/<Form ID>.deep.json` exports a Form with all its answers. Filters and processors can be added, as shown in the examples below.
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- `.data`: Include the forms that answer this questionnaire
+- `.dataFilter:modifiedAfter=2025-07-19T02:00:00%5C.000-05:00`: Files modified after 2 AM on July 19th, UTC -5. Do note that the `\` from a standard time stamp has been URL encoded to `%5C`. The shorter format `2025-07-19` is also supported, and interpreted as Midnight (time `T00:00:00.000`) in the server's timezone.
+- `.dataFilter:modifiedBefore=2025-07-26T02:00:00%5C.000-05:00`: Files modified before 2 AM on July 26th, UTC -5
+- `.labels`: Include the human readable version of answers, instead of the raw data (eg. `Never` instead of `0`)
```
```suggestion
- `.labels`: Use the human readable version of answers, instead of the raw data (eg. `Never` instead of `0`)
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+`/Questionnaires/OAIP.data.dataFilter:modifiedAfter=2025-07-19T02:00:00%5C.000-05:00.dataFilter:modifiedBefore=2025-07-26T02:00:00%5C.000-05:00.labels.formToSurveyLinks.dataFilter:status=SUBMITTED.csvIncludeFields:@survey=Survey.csvHeader:raw.questionnaireFilter:exclude=%252FQuestionnaires%252FOAIP%252Foaip_module1%252Foaip_visit_month.csv`
+- `/Questionnaires/OAIP`: Export the OAIP questionnaire
+- `.data`: Include the forms that answer this questionnaire
+- `.dataFilter:modifiedAfter=2025-07-19T02:00:00%5C.000-05:00`: Files modified after 2 AM on July 19th, UTC -5. Do note that the `\` from a standard time stamp has been URL encoded to `%5C`. The shorter format `2025-07-19` is also supported, and interpreted as Midnight (time `T00:00:00.000`) in the server's timezone.
```
```suggestion
- `.dataFilter:modifiedAfter=2025-07-19T02:00:00%5C.000-05:00`: Files modified after 2 AM on July 19th, UTC -5. Since the dot is used to separate the selectors, it must be escaped with `\`, which needs to be URL-encoded as `%5C`. The shorter format `2025-07-19` is also supported, and interpreted as Midnight (time `T00:00:00.000`) in the server's timezone.
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- `.dataFilter:modifiedBefore=2025-07-26T02:00:00%5C.000-05:00`: Files modified before 2 AM on July 26th, UTC -5
+- `.labels`: Include the human readable version of answers, instead of the raw data (eg. `Never` instead of `0`)
+- `.formToSurveyLinks`: Include the path to the relevant Survey Events form in the form data. This adds an `@survey` property, which is included in the export later
+- `.dataFilter:status=SUBMITTED`: Only include forms with the `SUBMITTED` status flag
```
Let's put the data filters together, and the survey related selectors together.
```suggestion
- `.dataFilter:status=SUBMITTED`: Only include forms with the `SUBMITTED` status flag
- `.labels`: Include the human readable version of answers, instead of the raw data (eg. `Never` instead of `0`)
- `.formToSurveyLinks`: Include the path to the relevant Survey Events form in the form data. This adds an `@survey` property, which is included in the export using the next line
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- `/Forms/<Form ID>.md` exports a Markdown-formatted view of a Form. Use `.txt` to export plain text instead.
+
+#### Example: Weekly OAIP Form Export
+`/Questionnaires/OAIP.data.dataFilter:modifiedAfter=2025-07-19T02:00:00%5C.000-05:00.dataFilter:modifiedBefore=2025-07-26T02:00:00%5C.000-05:00.labels.formToSurveyLinks.dataFilter:status=SUBMITTED.csvIncludeFields:@survey=Survey.csvHeader:raw.questionnaireFilter:exclude=%252FQuestionnaires%252FOAIP%252Foaip_module1%252Foaip_visit_month.csv`
```
```suggestion
`/Questionnaires/OAIP.data.dataFilter:modifiedAfter=2025-07-19T02:00:00%5C.000-05:00.dataFilter:modifiedBefore=2025-07-26T02:00:00%5C.000-05:00.dataFilter:status=SUBMITTED.labels.formToSurveyLinks.csvIncludeFields:@survey=Survey.csvHeader:raw.questionnaireFilter:exclude=%252FQuestionnaires%252FOAIP%252Foaip_module1%252Foaip_visit_month.csv`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- `.labels`: Include the human readable version of answers, instead of the raw data (eg. `Never` instead of `0`)
+- `.formToSurveyLinks`: Include the path to the relevant Survey Events form in the form data. This adds an `@survey` property, which is included in the export later
+- `.dataFilter:status=SUBMITTED`: Only include forms with the `SUBMITTED` status flag
+- `.csvIncludeFields:@survey=Survey`: Special instruction for the csv output format. Include the `@survey` property in the output column with the label `Survey`
```
```suggestion
- `.csvIncludeFields:@survey=Survey`: Special instruction for the csv output format. Include the `@survey` property in a column with the label `Survey`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+#### Example: Exporting all the data for a Visit as JSON
+`/Subjects/<MRN>/<Encounter ID>.data.deep.json`
+
+#### Example: Exporting a Patient and all the forms, from all their visits, that have been modified since the specified date
```
```suggestion
#### Example: Exporting a Patient and all the forms, from all their visits, that have been modified since the specified date, in a more compact JSON (`.bare`)
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+
+#### Example: Exporting a Patient and all the forms, from all their visits, that have been modified since the specified date
+`/Subjects/<MRN>.deep.data.dataFilter:modifiedAfter=2025-07-19.dataOption:descendantData=true.dataOption:formSelectors=deep%5C.bare.json`
+
```
```suggestion
- `.deep`: Include the Patient subject descendants, i.e. the visits
- `.data`: Include the forms for these subjects
- `dataFilter:modifiedAfter=2025-07-19`: Only include forms modified after the given date
- `.dataOption:descendantData=true`: Also include forms for the descendant subjects
- `dataOption:formSelectors=deep%5C.bare`: Apply a different set of selectors for serializing the forms. The separator dots must be escaped in order to not be considered as separators between the outer list of selectors.

```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- Processors and filters that expect a date to be provided can accept that date in a variety of formats. For example,
+    - `2025-01-01T02:00:00%5C.000-05:00`: A fully specified date, including date (Jan. 1), time (02:00:00), millisecond (.000), and timezone (-05:00)
+    - `2025-01-01T02:00`: A simplified datetime, interpretted using the server's timezone
+    - `2025-01-01`: A date, interpretted as midnight on that day in the server's timezone
```
```suggestion
    - `2025-01-01T02:00`: A simplified datetime, interpreted using the server's timezone
    - `2025-01-01`: A date, interpreted as midnight on that day in the server's timezone
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+#### Implementations:
+- **Description**: Simplify serialization for all resource types by removing all technical properties, renaming `jcr:created` to `created` and storing file attachments in a `content` property.
+- **Description**: Only runs on `Forms`.
+Simplify form serialization by only including the subject name, questionnaire and each sectionand question. Designed for use with `deep`, `-dereference` and `-identify`
```
```suggestion
Simplify form serialization by only including the subject name, questionnaire and each section and question. Designed for use with `deep`, `-dereference` and `-identify`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- **Description**: Only runs on `Forms`.
+Simplify form serialization by only including the subject name, questionnaire and each sectionand question. Designed for use with `deep`, `-dereference` and `-identify`
+- **Description**: Only runs on `Subjects`.
+Simplify subject serialization by only including simple labels for the name, type and parents of asubject. Should be used with `-dereference` and `-identify`
```
```suggestion
Simplify subject serialization by only including simple labels for the name, type and parents of a subject. Should be used with `-dereference` and `-identify`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+**Description**: Enable deep serialization, i.e. including the serialization of all descendant items. For example, serializing a Form without `deep` would only output the form's properties. With `deep`, all the answers are included, organized by sections.
+### dereference:
+**Description**: Dereference properties of type `REFERENCE`, `WEAKREFERENCE` and `PATH`: Instead of printing the internal UUID, serialize the referenced node.
+**isEnabledByDefault**: true
```
I would make this simpler:

```suggestion
**Enabled by default**
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+**isEnabledByDefault**: true
+### excludeDefaultProperties:
+**Description**: Exclude properties if their value is the default value or an obvious non-value:
+ - If a property's' value matches its JCR-defined default value
```
```suggestion
 - If a property's value matches its defined default value
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+**Description**: Exclude the contents of uploaded files from the serialization. By default, all uploaded files are excluded
+**isEnabledByDefault**: true
+#### Options:
+**exclude**: If this option is included, only exclude the specified file. `.excludeFiles:exclude=/Questionnaires/Path/To/QuestionWithFileAnswer`
```
```suggestion
**exclude**: If this option is specified, only exclude the specified files. Use it multiple times to exclude more than one file. `.excludeFiles:exclude=/Questionnaires/Path/To/QuestionWithFileAnswer`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+**Description**: Identify a node by including its `@path` and `@name` properties.
+**isEnabledByDefault**: true
+### importable:
+**Description**: Remove any properties from the output that would not be present in an importable questionnaire XML.This processor is intended to be run alongside the `deep` and `-identify` processor
```
```suggestion
**Description**: Export a file that is suitable for using as a source file to be imported, removing any properties that are tied to a running instance, such as unique ID and version control. This processor is intended to be run alongside the `deep` and `-identify` processor
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+### importable:
+**Description**: Remove any properties from the output that would not be present in an importable questionnaire XML.This processor is intended to be run alongside the `deep` and `-identify` processor
+### instanceCount:
+**Description**: Only runs on `Subjects`.
```
```suggestion
**Description**: Only runs on `SubjectTypes`.
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+Adds a label to answer option nodes.
+**isEnabledByDefault**: true
+- **Description**: Only runs on `Forms`.
+Add the human readable version of an answer to its serialization in the `displayedValue` propertyGet the human readable answer for text and number questions with options by outputting the answer option's label instead of the stored value.
```
Missing separator in this and the following lines, `propertyGet`.

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+Add the human readable version of an answer to its serialization in the `displayedValue` propertyGet the human readable answer for file questions by outputting the file name instead of the path.
+**isEnabledByDefault**: true
+- **Description**: Only runs on `Forms`.
+Add the human readable version of an answer to its serialization in the `displayedValue` propertyGet the human readable answer for file questions by outputting the file name instead of the path.
```
Why is this description present twice?

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+Add the human readable version of an answer to its serialization in the `displayedValue` propertyGet the human readable answer for boolean questions by outputting the labels specified in the question definition, e.g. 'Yes' or 'True' instead of '1'.
+**isEnabledByDefault**: true
+- **Description**: Only runs on `Forms`.
+Add the human readable version of an answer to its serialization in the `displayedValue` propertyGet the human readable answer for text and number questions with options by outputting the answer option's label instead of the stored value.
```
Reading through all of these, I think we should remove the first part and only mention the specific thing each of them does:

```suggestion
Get the human readable answer for text and number questions with options by outputting the answer option's label instead of the stored value.
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+It is not possible to include a descendant of an excluded item
+#### Options:
+**include**: A path to an included question or section. `answerFilter:include=/Questionnaires/Path/To/Question`
+**exclude**: A path to an excluded question or section. `answerFilter:exclude=/Questionnaires/Path/To/Question`
```
```suggestion
**include**: A path to an included question or section. `questionnaireFilter:include=/Questionnaires/Path/To/Question`
**exclude**: A path to an excluded question or section. `questionnaireFilter:exclude=/Questionnaires/Path/To/Question`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+- **Description**: Only runs on `Questionnaires`.
+Simplify questionnaire serialization by removing child jcr properties.
+- **Description**: Only runs on `Forms`.
+Simplify form serialization by removing unnecessary properties and children.Removes child jcr properties, removes extra properties from answers and cleans up subject types
```
```suggestion
Simplify form serialization by removing unnecessary properties and children. Removes child jcr properties, removes extra properties from answers and cleans up subject types
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+### clinicNot:
+**Description**: Exclude forms that belong to a user that has a 'Visit information' form for the specified clinic. If included multiple times, this excludes forms belonging to any of the specified clinics.
+### createdAfter:
+**Description**: Only show results that were created after the requested datetime. e.g. `.dataFilter:createdAfter=2025-01-01T06:00:00%5C.000-05:00` for forms created after January 1, 2025 at 6 AM in the time zone UTC-5
```
```suggestion
**Description**: Only show results that were created on or after the requested datetime. e.g. `.dataFilter:createdAfter=2025-01-01T06:00:00%5C.000-05:00` for forms created after January 1, 2025 at 6 AM in the time zone UTC-5
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+
+## Filters:
+### clinic:
+**Description**: Only show forms that belong to a user that has a 'Visit information' form for the specified clinic. If included multiple times, this includes forms belonging to any of the specified clinics.
```
I think all the filters should have an example.

```suggestion
**Description**: Only show forms that belong to a user that has a 'Visit information' form for the specified clinic. If included multiple times, this includes forms belonging to any of the specified clinics. e.g. `.clinic=PMH-YVM`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+### createdBefore:
+**Description**: Only show results that were created before the requested datetime. e.g. `.dataFilter:createdBefore=2025-01-01T06:00:00%5C.000-05:00` for forms created before January 1, 2025 at 6 AM in the time zone UTC-5
+### createdBy:
+**Description**: Only show results that were created by the specified user.
```
```suggestion
**Description**: Only show results that were created by the specified user. e.g. `.createdBy=admin`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+### notModifiedBy:
+**Description**: Only show results that were last modified by anby other other than the specified user.
+### status:
+**Description**: Only show results that have the specified status flag.
```
```suggestion
**Description**: Only show results that have the specified status flag. e.g. `.status=SUBMITTED`
```

sdumitriu commented on modules/utils/Serialization Readme.md:
```diff
@@ -0,0 +1,242 @@
+### status:
+**Description**: Only show results that have the specified status flag.
+### statusNot:
+**Description**: Only show results that do not have the specified status flag.
```
```suggestion
**Description**: Only show results that do not have the specified status flag. e.g. `statusNot=INCOMPLETE`
```

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/api/SelectorServlet.java:
```diff
@@ -0,0 +1,256 @@
+ * @version $Id$
+ */
+@Component(service = { Servlet.class })
+@SlingServletResourceTypes(resourceTypes = { "cards/QuestionnairesHomepage" }, extensions = {
```
```suggestion
@SlingServletResourceTypes(resourceTypes = { "cards/Homepage" }, extensions = {
```

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/api/SelectorServlet.java:
```diff
@@ -0,0 +1,256 @@
+
+/**
+ * A servlet that lists the available {@code ResourceJsonProcessor}s and {@code DataFilter}s
+ * which can be used as selectors for questionnaire data exports.
```
```suggestion
 * which can be used as selectors for data serialization or exports.
```

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
- Update descriptions based on feedback

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
Update code for generating endpoint output
- Improve markdown generation and formating to avoid Json intermediate step
- Add details for the ResourceCSVProcessor
- Improve markdown formatting
- Comment code

acrowthe committed:
CARDS-2723: Create an endpoint to dump data processors and filters
Improve selector descriptions
- End descriptions with periods where applicable
- Clean up vocabulary answer label processor description
- Improve appearance of empty string in excludeDefaultProperties description
- Update serialization document with newly generated endpoint output
- Fix a couple other typos

sdumitriu committed:
[misc] Fix typos and wrong copy/paste
